### PR TITLE
JW-1242: refactorings in vr_packet functions in vr_nbl.c

### PR DESCRIPTION
**Pull request by sodar**
- `win_get_packet` renamed to `AllocateVrPacketForNetBufferList`
    - also some minor refactorings
- `win_allocate_packet` renamed to `AllocateVrPacket`
- `win_free_packet` renamed to `FreeVrPacket`

### CODE REVIEW

**Comment by sodar** [comment]
Retest this please

File: windows/vr_host.c [hunk 1]
```diff
@@ -261,47 +266,40 @@ win_preset(struct vr_packet *pkt)
...
+    if (status != NDIS_STATUS_SUCCESS)
+        goto cleanupNbl;
 
-    npkt->vp_net_buffer_list = nbl;
+    struct vr_packet *newPacket = GetVrPacketFromNetBufferList(newNbl);
+    ASSERT(newPacket != NULL);
 
-    npkt->vp_cpu = (unsigned char)win_get_cpu();
+    RtlCopyMemory(newPacket, originalPacket, sizeof(*originalPacket));
+    newPacket->vp_cpu = (unsigned char)KeGetCurrentProcessorNumberEx(NULL);
```
Comment (jablonskim) [inline windows/vr_host.c]: Please remove (?)

File: windows/vr_nbl.c [hunk 2]
```diff
@@ -320,10 +294,10 @@ win_allocate_packet(void *buffer, unsigned int size)
 }
 
 void
-win_free_packet(struct vr_packet *pkt)
+FreeVrPacket(struct vr_packet *pkt)
 {
     ASSERT(pkt != NULL);
-    free_associated_nbl(pkt);
+    FreeNetBufferList(pkt->vp_net_buffer_list);
```
Comment (jablonskim) [inline windows/vr_nbl.c]: Maybe add ASSERT(pkt->vp_net_buffer_list != NULL)?

File: windows/vr_host.c [hunk 3]
```diff
@@ -922,46 +940,45 @@ win_soft_reset(struct vrouter *router)
...
-            RtlAnsiStringToUnicodeString(&unicode_name, &ansi_name, TRUE);
-
-            if (memcmp(unicode_name.Buffer, element->NicName.String, (element->NicName.Length < unicode_name.Length ? element->NicName.Length : unicode_name.Length)) == 0)
-            {
-                vif->vif_port = element->PortId;
-                vif->vif_nic = element->NicIndex;
+        }  else if (element->NicType == NdisSwitchNicTypeEmulated || element->NicType == NdisSwitchNicTypeSynthetic) {
+            ANSI_STRING ansiName;
+            ansiName.Buffer = vifr->vifr_if_guid;
+            ansiName.Length = vifr->vifr_if_guid_size + 1; // For NULL character
```
Comment (krdln) [inline windows/vr_host.c]: https://msdn.microsoft.com/en-us/library/windows/hardware/ff540605(v=vs.85).aspx

> If the string is null-terminated, Length does not include the terminating NULL.

Is that +1 used here then for the `memcmp` not to return false positives when names are different, but substrings?

File: windows/vr_host.c [hunk 4]
```diff
@@ -922,46 +940,45 @@ win_soft_reset(struct vrouter *router)
...
+            ansiName.MaximumLength = vifr->vifr_if_guid_size + 1; // For NULL character
 
-                RtlFreeUnicodeString(&unicode_name);
+            UNICODE_STRING unicodeName;
+            RtlAnsiStringToUnicodeString(&unicodeName, &ansiName, TRUE);
+            ULONG length = (element->NicName.Length < unicodeName.Length ? element->NicName.Length : unicodeName.Length);
 
-                break;
+            if (memcmp(unicodeName.Buffer, element->NicName.String, length) == 0) {
+                vif->vif_port = element->PortId;
```
Comment (krdln) [inline windows/vr_host.c]: This `memcmp` actually checks if `element->NicName` is a **prefix** of `unicodeName`. For example, the condition will be true for `NicName == ""`. We can check `unicodeName.length == NicName.Length` or just use [RtlEqualUnicodeString](https://msdn.microsoft.com/en-us/library/windows/hardware/ff561854(v=vs.85).aspx) (that'd require temporary `UNICODE_STRING` for `NicName`).

Comment (sodar) [reply windows/vr_host.c]: I think using `RtlEqualUnicodeString` is the way to go here.

**Comment by krdln** [comment]
Retest this please

**Comment by krdln** [comment]
Retest this please

File: windows/vr_nbl.c [hunk 5]
```diff
@@ -482,6 +495,29 @@ FilterSendNetBufferListsComplete(
...
+        fwdDetail = NET_BUFFER_LIST_SWITCH_FORWARDING_DETAIL(current);
+        if (fwdDetail->NativeForwardingRequired) {
+            CompleteReceivedNetBufferList(current);
+            DbgPrint("!!! Internal packed found in FilterSendNetBufferListComplete\r\n");
+        } else {
+            struct vr_packet *pkt = GetVrPacketFromNetBufferList(current);
+            unsigned char thisCore = KeGetCurrentProcessorNumberEx(NULL);
+            unsigned char oldCore = pkt->vp_cpu;
+            if (thisCore != oldCore) {
+                DbgPrint("!!! Complete -- packet switched cores: %hhd -> %hhd\r\n", oldCore, thisCore);
```
Comment (sodar) [inline windows/vr_nbl.c]: I think we do not need this core checking now.

File: windows/vr_nbl.c [hunk 6]
```diff
@@ -482,6 +495,29 @@ FilterSendNetBufferListsComplete(
...
+            }
+
+            // TODO proper msg
+            ASSERTMSG("Completed NBLS should already be marked as garbage ", pkt->vp_net_buffer_list == NULL);
```
Comment (sodar) [inline windows/vr_nbl.c]: I this `ASSERTMSG` is this proper message, please remove the comment

File: windows/vr_nbl.c [hunk 7]
```diff
@@ -482,6 +495,29 @@ FilterSendNetBufferListsComplete(
...
+
+            if (IS_NBL_OWNED(current) && IS_NBL_CLONE(current)) {
+                PNET_BUFFER_LIST parent = current->ParentNetBufferList;
+                struct vr_packet *parent_pkt = GetVrPacketFromNetBufferList(parent);
+                ASSERTMSG("Completed NBLS' parents should already be marked as garbage ", parent_pkt->vp_net_buffer_list == NULL);
```
Comment (sodar) [inline windows/vr_nbl.c]: As per our discussion, this `if` and `ASSERTMSG` can be removed.

**Comment by krdln** [comment]
Retest this please

**Pull request merged**